File: modules/video-pose-analysis/android/src/main/java/com/divotgolf/videoposeanalysis/VideoPoseAnalysisModule.kt
package com.divotgolf.videoposeanalysis

import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class VideoPoseAnalysisModule : Module() {
    companion object {
        private const val TAG = "VideoPoseAnalysis"
    }

    @Volatile
    private var currentPipeline: VideoPosePipeline? = null

    override fun definition() = ModuleDefinition {
        Name("VideoPoseAnalysis")

        Events("onPoseAnalysisProgress")

        AsyncFunction("analyzeVideo") { filePath: String, clipId: String ->
            Log.d(TAG, "analyzeVideo called: clipId=$clipId, path=$filePath")

            val context = appContext.reactContext
                ?: throw IllegalStateException("React context not available")

            val pipeline = VideoPosePipeline(context, filePath)
            currentPipeline = pipeline

            pipeline.onProgress = { progress, currentFrame, totalFrames ->
                sendEvent("onPoseAnalysisProgress", mapOf(
                    "progress" to progress,
                    "currentFrame" to currentFrame,
                    "totalFrames" to totalFrames,
                ))
            }

            val result = pipeline.run(clipId)
            currentPipeline = null
            result
        }

        Function("cancelAnalysis") {
            Log.d(TAG, "cancelAnalysis called")
            currentPipeline?.isCancelled = true
        }

        OnDestroy {
            Log.d(TAG, "OnDestroy: cancelling any running analysis")
            currentPipeline?.isCancelled = true
            currentPipeline = null
        }
    }
}

File: modules/video-pose-analysis/ios/VideoPoseAnalysisModule.swift
import ExpoModulesCore
import os.log

private let logger = Logger(subsystem: "com.divotgolf.videoposeanalysis", category: "VideoPoseAnalysisModule")

public class VideoPoseAnalysisModule: Module {
    private var currentPipeline: VideoPosePipeline?

    public func definition() -> ModuleDefinition {
        Name("VideoPoseAnalysis")

        Events("onPoseAnalysisProgress")

        AsyncFunction("analyzeVideo") { (filePath: String, clipId: String) -> [String: Any] in
            logger.info("analyzeVideo called: clipId=\(clipId), path=\(filePath)")

            let url: URL
            if filePath.hasPrefix("file://") {
                guard let parsed = URL(string: filePath) else {
                    throw VideoPoseError.noVideoTrack
                }
                url = parsed
            } else {
                url = URL(fileURLWithPath: filePath)
            }

            let pipeline = try VideoPosePipeline(url: url)
            self.currentPipeline = pipeline

            pipeline.onProgress = { [weak self] progress, currentFrame, totalFrames in
                DispatchQueue.main.async {
                    self?.sendEvent("onPoseAnalysisProgress", [
                        "progress": progress,
                        "currentFrame": currentFrame,
                        "totalFrames": totalFrames,
                    ])
                }
            }

            let result = try pipeline.run(clipId: clipId)
            self.currentPipeline = nil
            return result
        }

        Function("cancelAnalysis") {
            logger.info("cancelAnalysis called")
            self.currentPipeline?.isCancelled = true
        }

        OnDestroy {
            logger.info("OnDestroy: cancelling any running analysis")
            self.currentPipeline?.isCancelled = true
            self.currentPipeline = nil
        }
    }
}
